### PR TITLE
Allow specifying `errorIdentifier` to `RuleError`

### PR DESCRIPTION
**Pull request by ruudk**
Ultimately every error in PHPStan should have its own unique identifier so that its
possible to group by error type and ignore errors based on identifier.

https://github.com/phpstan/phpstan/issues/1686#issuecomment-532996159
https://github.com/phpstan/phpstan/issues/3065

### CODE REVIEW

**Comment by spaze** [comment]
Oh nice, thanks! I always wanted to start using `RuleErrorBuilder` but never got around to it. I'll check the whole PR later.

File: src/DisallowedCall.php [hunk 1]
```diff
@@ -55,6 +59,7 @@ public function __construct(string $call, ?string $message, array $allowIn, arra
 		$this->allowParamsAnywhere = $allowParamsAnywhere;
 		$this->allowExceptParamsInAllowed = $allowExceptParamsInAllowed;
 		$this->allowExceptParams = $allowExceptParams;
+		$this->identifier = (string) $identifier;
```
Comment (spaze) [inline src/DisallowedCall.php:62]: The method only accepts `string`, you can remove the typecast.

File: src/DisallowedCall.php [hunk 2]
```diff
@@ -124,6 +129,15 @@ public function getAllowExceptParams(): array
 	}
 
 
+	/**
+	 * @return string
```
Comment (spaze) [inline src/DisallowedCall.php:133]: Can remove the doctype as the return type is in the method signature already.

File: src/DisallowedConstant.php [hunk 3]
```diff
@@ -51,4 +56,13 @@ public function getAllowIn(): array
 		return $this->allowIn;
 	}
 
+
+	/**
+	 * @return string
```
Comment (spaze) [inline src/DisallowedConstant.php:61]: Same here, can drop the doctype.

File: src/DisallowedNamespace.php [hunk 4]
```diff
@@ -49,4 +54,13 @@ public function getAllowIn(): array
 		return $this->allowIn;
 	}
 
+
+	/**
+	 * @return string
```
Comment (spaze) [inline src/DisallowedNamespace.php:59]: And here as well, can remove the doctype :-)

File: phpstan.neon [hunk 5]
```diff
@@ -3,7 +3,7 @@ parameters:
...
+		ForbiddenCallsConfig: 'array<array{function?:string, method?:string, message?:string, allowIn?:string[], allowInFunctions?:string[], allowInMethods?:string[], allowParamsInAllowed?:array<integer, integer|boolean|string>, allowParamsInAllowedAnyValue?:array<integer, integer>, allowParamsAnywhere?:array<integer, integer|boolean|string>, allowParamsAnywhereAnyValue?:array<integer, integer>, allowExceptParamsInAllowed?:array<integer, integer|boolean|string>, allowExceptParams?:array<integer, integer|boolean|string>, allowExceptCaseInsensitiveParams?:array<integer, integer|boolean|string>, identifier?: string}>'
```
Comment (spaze) [inline phpstan.neon:6]: Thinking about the `identifier` keyword, seems too generic. Maybe "errorIdentifier"? "errorId"?

Comment (spaze) [reply phpstan.neon:6]: Also please remove the space after `:` (yeah, I know 🤭)

Comment (spaze) [reply phpstan.neon:6]: Also also, can you please add a line to README that would describe the config keyword (and make the support official and documented)?

**Review by spaze — CHANGES_REQUESTED** [review]
Thanks for the PR. Some smaller things I'd like (youl to address 😀 I've also merged the previous PR so you can rebase this one.

**Comment by ruudk** [comment]
@spaze I have no idea how to solve the phpcs issues. They conflict with each other. If I run `composer run cs-fix` it cannot fix it. What to do?

**Comment by spaze** [comment]
> @spaze I have no idea how to solve the phpcs issues. They conflict with each other. If I run `composer run cs-fix` it cannot fix it. What to do?

Oh what a mess! :-D Couldn't hit the problem first but then I ran `composer update` and voila, shit. Seems that a [new code sniffer version](https://github.com/squizlabs/PHP_CodeSniffer/releases/tag/3.6.2) released just yesterday "Added new PSR12.Classes.OpeningBraceSpace sniff to enforce this" (https://github.com/squizlabs/PHP_CodeSniffer/issues/3437).

I'll update the coding style to remove the new requirement. Sorry for the trouble and thanks for hitting it before I do :-)

**Comment by spaze** [comment]
@ruudk all green now again :-) As an extra I've also added PHP 8.1 tests and rebased this branch. Let me know once the PR is ready, I'll re-review or something. Currently sort of busy so please allow a day or 2 (or 5). Thanks!

File: src/Usages/ClassConstantUsages.php [hunk 6]
```diff
@@ -78,11 +80,11 @@ public function processNode(Node $node, Scope $scope): array
 		} else {
 			if ($usedOnType->hasConstant($constant)->no()) {
 				return [
-					sprintf(
+					RuleErrorBuilder::message(sprintf(
 						'Cannot access constant %s on %s',
 						$constant,
 						$usedOnType->describe(VerbosityLevel::getRecommendedLevelByType($usedOnType))
-					),
+					))->build(),
```
Comment (spaze) [inline src/Usages/ClassConstantUsages.php:87]: Note to self: no need for `->identifier(...)` here because the error is not a *disallowed call* error.

**Pull request merged**